src/vm/genericdict.cpp

```diff
@@ -109,183 +109,333 @@ DictionaryLayout::GetFirstDictionaryBucketSize(
+        PRECONDITION(CheckPointer(pMD));
+        PRECONDITION(CheckPointer(pAllocator));
+        PRECONDITION(CheckPointer(pResult));
+        PRECONDITION(pMD->HasMethodInstantiation());
+    }
+    CONTRACTL_END
+
+    DWORD cbSig = -1;
+    pSig = pSigBuilder != NULL ? (BYTE*)pSigBuilder->GetSignature(&cbSig) : pSig;
+    if (FindTokenWorker(pAllocator, pMD->GetNumGenericMethodArgs(), pMD->GetDictionaryLayout_Unsafe(), pSigBuilder, pSig, cbSig, nFirstOffset, signatureSource, pResult, pSlotOut, 0, FALSE))
+        return TRUE;
+
+    CrstHolder ch(&SystemDomain::SystemModule()->m_DictionaryCrst);
+    {
+        // Try again under lock in case another thread already expanded the dictionaries or filled an empty slot
+        if (FindTokenWorker(pAllocator, pMD->GetNumGenericMethodArgs(), pMD->GetDictionaryLayout_Unsafe(), pSigBuilder, pSig, cbSig, nFirstOffset, signatureSource, pResult, pSlotOut, *pSlotOut, TRUE))
+            return TRUE;
+
+#ifndef CROSSGEN_COMPILE
+        DictionaryLayout* pOldLayout = pMD->GetDictionaryLayout_Unsafe();
+        DictionaryLayout* pNewLayout = ExpandDictionaryLayout(pAllocator, pOldLayout, pMD->GetNumGenericMethodArgs(), pSigBuilder, pSig, nFirstOffset, signatureSource, pResult, pSlotOut);
+        if (pNewLayout == NULL)
+        {
+            pResult->signature = pSigBuilder == NULL ? pSig : CreateSignatureWithSlotData(pSigBuilder, pAllocator, 0);
+            return FALSE;
+        }
+
+        // First, expand the PerInstInfo dictionaries on methods that were using the dictionary layout that just got expanded, and expand their slots
+        pMD->GetModule()->ExpandMethodDictionaries_Locked(pMD, pOldLayout, pNewLayout);
 
-    return FindTokenWorker(pAllocator, numGenericArgs, pDictLayout, pResult, NULL, signature, -1, nFirstOffset, signatureSource, pSlotOut);
+        // Ensure no other thread uses old dictionary pointers
+        FlushProcessWriteBuffers();
+
+        // Finally, update the dictionary layout pointer after all dictionaries of instantiated methods have expanded, so that subsequent calls to 
+        // DictionaryLayout::FindToken can use this. It is important to update the dictionary layout at the very last step, otherwise some other threads
+        // can start using newly added dictionary layout slots on methods where the PerInstInfo hasn't expanded yet, and cause runtime failures.
+        pMD->AsInstantiatedMethodDesc()->IMD_SetDictionaryLayout(pNewLayout);
+
+        return TRUE;
+#else
+        pResult->signature = pSigBuilder == NULL ? pSig : CreateSignatureWithSlotData(pSigBuilder, pAllocator, 0);
+        return FALSE;
+#endif
+    }
+}
+
+/* static */
+PVOID DictionaryLayout::CreateSignatureWithSlotData(SigBuilder* pSigBuilder, LoaderAllocator* pAllocator, WORD slot)
+{
+    CONTRACTL
+    {
+        STANDARD_VM_CHECK;
+        PRECONDITION(CheckPointer(pSigBuilder) && CheckPointer(pAllocator));
+    }
+    CONTRACTL_END
+
+    pSigBuilder->AppendData(slot);
```

Comment:
Do we still need to store the slot in the signature in the new scheme - is it needed by anything?